### PR TITLE
Clarify Locker.Locked(), and switch some locks to read-only

### PR DESCRIPTION
**Pull request by nalind**
Clarify that our `Locker.Locked()` checks if the calling process is holding a write lock, since that's what we care about in the places where we call it.  Change a number of places where `Store` takes a lock but doesn't expect to be modifying on-disk data to have it take a read-only lock instead.

### CODE REVIEW

**Comment by vrothberg** [comment]
LGTM :+1: 
Would you open test-PRs for CRI-O and Buildah to make sure there's no unexpected side effect?

**Comment by rhatdan** [comment]
LGTM

**Comment by rhatdan** [comment]
@nalind Is this ready merge.

File: layers.go [hunk 1]
```diff
@@ -718,6 +774,14 @@ func (r *layerStore) Unmount(id string, force bool) (bool, error) {
...
 		return false, errors.Wrapf(ErrStoreIsReadOnly, "not allowed to update mount locations for layers at %q", r.mountspath())
 	}
+	r.mountsLockfile.Lock()
+	defer r.mountsLockfile.Unlock()
+	if modified, err := r.mountsLockfile.Modified(); modified || err != nil {
+		if err = r.loadMounts(); err != nil {
+			return false, err
+		}
+	}
+	defer r.mountsLockfile.Touch()
```
Comment (vrothberg) [inline layers.go]: In addition to the other panic()s in the lockfile, we might add another one to `Touch()` if the lock isn't locked and writable.

Comment (nalind) [reply layers.go]: Hmm, I like that idea.

**Comment by nalind** [comment]
I didn't clock it, but I did count the number of write locks we took during a couple of `buildah images` runs, and while https://github.com/containers/buildah/pull/1352 helped for the `--all` case, this reduced things by more than that one did.

The additional lock is new since I first opened the PR, so if those additions also look good, I'm okay with merging it.

**Comment by nalind** [comment]
https://github.com/kubernetes-sigs/cri-o/pull/2089 is currently failing because the new "attempted to acquire one lock type while holding another" check is being tripped when multiple cri-o threads are interacting with the locks.  We don't want to trigger an error in that case, though.

**Comment by vrothberg** [comment]
> [kubernetes-sigs/cri-o#2089](https://github.com/kubernetes-sigs/cri-o/pull/2089) is currently failing because the new "attempted to acquire one lock type while holding another" check is being tripped when multiple cri-o threads are interacting with the locks. We don't want to trigger an error in that case, though.

This makes sense to me. I just had a look at `func (l *lockfile) lock(l_type int16)` and spotted another deadlock when we let this pass. I will try to fix this in a separate PR where you can pick the commit from.

**Comment by vrothberg** [comment]
Here's the fix: https://github.com/containers/storage/pull/290

**Comment by nalind** [comment]
Cherry picked #290.

**Comment by nalind** [comment]
I'm concerned that we may hit problems with writer starvation, but we may have to see.

**Comment by rhatdan** [comment]
Passes tests on CRI-O.  Merging.

**Pull request merged**